### PR TITLE
Introducing BatchBlock

### PR DESCRIPTION
**Pull request by marcromeyn**
### Goals :soccer:
`BatchBlock` will be used inside the `Model` to create the `Batch` object. It's also useful for things like masking & padding.

### CODE REVIEW

File: tests/unit/torch/test_block.py [hunk 1]
```diff
@@ -377,3 +377,58 @@ def forward(self, inputs: Dict[str, torch.Tensor]) -> Tuple[torch.Tensor, torch.
...
+    def test_forward_with_tuple(self):
+        feat, target = torch.tensor([1, 2]), torch.tensor([3, 4])
+        outputs = module_utils.module_test(mm.BatchBlock(), feat, targets=target)
+
+        assert isinstance(outputs, mm.Batch)
+        assert torch.equal(outputs.feature(), feat)
+        assert torch.equal(outputs.target(), target)
+
+    def test_forward_exception(self):
+        with pytest.raises(RuntimeError):
```
Comment (oliverholworthy) [inline tests/unit/torch/test_block.py:404]: I noticed that in some places we use the  `match=` argument of `pytest.raises` to specify which error we're expecting. I think this is a good pattern to adopt and will make things more robust. Since this could start failing from a different place accidentally and still pass the test, changing what the test is checking.

File: tests/unit/torch/test_batch.py [hunk 2]
```diff
@@ -186,6 +186,69 @@ def test_device(self):
...
+        )
+
+        # with inputs
+        input_batch = Batch(
+            {"feature2": torch.tensor([11, 12])}, {"target1": torch.tensor([13, 14])}, sequences
+        )
+        result = batch.flatten_as_dict(input_batch)
+        assert len(result) == 9  # input keys are considered
+        assert (
+            len([k for k in result if k.startswith("inputs.")]) == 4
```
Comment (sararb) [inline tests/unit/torch/test_batch.py:217]: what is the difference between `inputs.` and `features.` keys ?

File: tests/unit/torch/test_block.py [hunk 3]
```diff
@@ -377,3 +377,60 @@ def forward(self, inputs: Dict[str, torch.Tensor]) -> Tuple[torch.Tensor, torch.
...
+        feat, target = torch.tensor([1, 2]), torch.tensor([3, 4])
+        outputs = module_utils.module_test(mm.BatchBlock(mm.BatchBlock()), feat, targets=target)
+
+        assert isinstance(outputs, mm.Batch)
+        assert torch.equal(outputs.feature(), feat)
+        assert torch.equal(outputs.target(), target)
+
+    def test_in_parallel(self):
+        feat, target = torch.tensor([1, 2]), torch.tensor([3, 4])
+        outputs = module_utils.module_test(
```
Comment (sararb) [inline tests/unit/torch/test_block.py:419]: +1 to how we can now create different preprocessing blocks for each subset of the inputs! I love it!!

**Review by sararb — APPROVED** [review]
The PR looks good to me! I've just left some minor comments/questions

File: merlin/models/torch/batch.py [hunk 4]
```diff
@@ -277,6 +295,106 @@ def target(self, name: str = "default") -> torch.Tensor:
...
+            Another Batch object to include in the flattening process. The keys from the input
+            batch are also added with a prefix of "inputs.", by default None
+
+        Returns
+        -------
+        Dict[str, torch.Tensor]
+            A dictionary containing all the flattened features, targets, and sequences.
+        """
+        flat_dict: Dict[str, torch.Tensor] = self._flatten()
+        dummy_tensor = torch.tensor(0)
```
[the remark this answers is not in the document]
Comment (marcromeyn) [reply merlin/models/torch/batch.py:319]: We never use it, but we need to have it in order to keep the type of the `Dict[str, torch.Tensor]`. We could store the original value but that might take more memory, so that's why I added the dummy

File: merlin/models/torch/models/base.py [hunk 5]
```diff
@@ -128,9 +140,11 @@ def forward(
         self, inputs: Union[torch.Tensor, Dict[str, torch.Tensor]], batch: Optional[Batch] = None
     ):
         """Performs a forward pass through the model."""
+        _batch = self.pre(inputs, batch=batch)
+
         outputs = inputs
         for block in self.values:
```
[the remark this answers is not in the document]
Comment (marcromeyn) [reply merlin/models/torch/models/base.py:146]: Good point! Lemme update that
